rsconnect/bundle.py

```diff
@@ -1241,6 +1305,59 @@ def write_notebook_manifest_json(
     return exists(join(directory, environment.filename))
 
 
+def write_voila_manifest_json(
+    path: str,
+    entrypoint: str,
+    environment: Environment,
+    app_mode: AppMode = AppModes.JUPYTER_VOILA,
+    extra_files: typing.List[str] = None,
+    excludes: typing.List[str] = None,
+    force_generate: bool = True,
+    image: str = None,
+) -> bool:
+    """
+    Creates and writes a manifest.json file for the given path.
+
+    :param path: the file, or the directory containing the files to deploy.
+    :param entry_point: the main entry point for the API.
+    :param environment: the Python environment to start with.  This should be what's
+    returned by the inspect_environment() function.
+    :param app_mode: the application mode to assume.  If this is None, the extension
+    portion of the entry point file name will be used to derive one. Previous default = None.
+    :param extra_files: any extra files that should be included in the manifest. Previous default = None.
+    :param excludes: a sequence of glob patterns that will exclude matched files.
+    :param force_generate: bool indicating whether to force generate manifest and related environment files.
+    :param image: the optional docker image to be specified for off-host execution. Default = None.
+    :return: whether the manifest was written.
+    """
+    mimetypes.add_type("text/ipynb", ".ipynb")
+    extra_files = list(extra_files) if extra_files else []
+    excludes = list(excludes) if excludes else []
+
+    entrypoint = entrypoint or infer_entrypoint(path=path, mimetype="text/ipynb")
+    manifest_data = make_source_manifest(AppModes.JUPYTER_VOILA, environment, entrypoint, None, image)
+    base_dir = dirname(entrypoint)
+
+    if isfile(path):
+        validate_file_is_notebook(entrypoint)
+        manifest_add_file(manifest_data, entrypoint, base_dir)
+
+    # handle environment files
+    if not exists(join(base_dir, environment.filename)) or force_generate:
+        write_environment_file(environment, base_dir)
+    manifest_add_buffer(manifest_data, environment.filename, environment.contents)
+
+    excludes.extend(["manifest.json"])
+    file_list = create_file_list(path, extra_files, excludes)
+    for rel_path in file_list:
+        manifest_add_file(manifest_data, rel_path, path)
```

Comment:
Up to this point, this function appears identical to `make_voila_bundle`, and we need their behaviors to be identical so that deploying now or writing a manifest for later produce the same bundle. Can you factor out the common section so there is only one version of it?

Reply:
To make this happen I factored out data structures for Manifest and Bundle. This made data passing between functions easier and let us avoid the problem of doing unnecessary IO operations for disparate manifest and bundle related function calls.